scripts/codegen/modules/devicedeclare.py

```diff
@@ -0,0 +1,391 @@
+                         device_prios,
+                         device_api,
+                         device_info):
+    devices_declared = []
+    for i, device_config in enumerate(device_configs):
+        driver_name = driver_names[i]
+        if isinstance(device_inits, str):
+            device_init = device_inits
+        else:
+            try:
+                device_init = device_inits[i]
+            except:
+                device_init = device_inits
+        if isinstance(device_levels, str):
+            device_level = device_levels
+        else:
+            try:
+                device_level = device_levels[i]
+            except:
+                device_level = device_levels
+        if isinstance(device_prios, str):
+            device_prio = device_prios
+        else:
+            try:
+                device_prio = device_prios[i]
+            except:
+                device_prio = device_prios
+
+        device_declared = device_declare_single(device_config,
+                                                driver_name,
+                                                device_init,
+                                                device_level,
+                                                device_prio,
+                                                device_api,
+                                                device_info)
+        devices_declared.append(device_declared)
+
+    if True not in devices_declared:
+        err = "No active device found for {} = {} and {}.".format(
+            ', '.join(device_configs), ', '.join(devices_declared),
+            ', '.join(driver_names))
+        codegen.log(err)
+        raise codegen.Error(err)
+
+
+def _device_generate_struct(type_of_struct, _struct):
+    struct = ""
+
+    # convert _struct into a list. Struct might have only one element
+    if type(_struct) is str:
+        _struct = [_struct]
+    else:
+        _struct = list(_struct)
+
+    if type_of_struct == 'config':
+        struct += 'static const struct {} ${{device-config-info}} = {{\n'.format(_struct[0])
+    elif type_of_struct == 'data':
+        struct += 'static struct {} ${{device-data}} = {{\n'.format(_struct[0])
+    else:
+        msg("Not expected")
+
+    if len(_struct) > 1:
+        struct += _struct[1]
+
+    struct += '};\n\n'
+    return struct
+
+
+def _device_generate_irq_bootstrap(irq_names, irq_flag, irq_func):
+    irq_bootstrap_info = \
+        '#ifdef {}\n'.format(irq_flag) + \
+        'DEVICE_DECLARE(${device-name});\n' + \
+        'static void ${device-config-irq}(struct device *dev)\n' + \
+        '{\n'
+    for irq_name in irq_names:
+        irq_num = '${{interrupts/{}/irq}}'.format(irq_name)
+        irq_prio = '${{interrupts/{}/priority}}'.format(irq_name)
+        irq_bootstrap_info += \
+            '\tIRQ_CONNECT({},\n'.format(irq_num) + \
+            '\t\t{},\n'.format(irq_prio) + \
+            '\t\t{}_{},\n'.format(irq_func, irq_name) + \
```

Comment:
> Would be good to identify this and keep using 'driver_isr'

This adds complexity for a special case of a function name that is not visible in the source code. It is an  autogenerated function name that you only see in the generated code (and the linker listing).

Can you elaborate the benefit.

Reply:
>  It is an autogenerated function name that you only see in the generated code (and the linker listing).

This is not correct. isr function name is defined in core driver code.
Then, why should user define a 'spi_stm32_isr_0' (instead of 'spi_stm32_isr') function, if there is only one possible IRQ for his driver?

Reply:
To get rid of guessing. Please provide patch/ pull request that fits to the intention of your API.

Reply:
Had something working as follows:
```
-            '\t\t{},\n'.format(irq_prio) + \
-            '\t\t{}_{},\n'.format(irq_func, irq_name) + \
+            '\t\t{},\n'.format(irq_prio)
+        if len(irq_names) == 1 and irq_names[0] == '0':
+            # Only one irq and no name associated. Keep it simple name
+            irq_bootstrap_info += \
+                '\t\t{},\n'.format(irq_func, irq_name)
+        else:
+            irq_bootstrap_info += \
+                '\t\t{}_{},\n'.format(irq_func, irq_name)
+        irq_bootstrap_info += \
```
Maybe not the cleanest solution though.